pkg/apiserver/logsearch/router.go

```diff
@@ -0,0 +1,353 @@
+)
+
+type Service struct {
+	config    *config.Config
+	db        *dbstore.DB
+	scheduler *Scheduler
+}
+
+var logsSavePath string
+
+func NewService(config *config.Config, db *dbstore.DB) *Service {
+	logsSavePath = path.Join(config.DataDir, "logs")
+	os.MkdirAll(logsSavePath, 0777) //nolint:errcheck
+
+	autoMigrate(db)
+	cleanRunningTasks(db)
+
+	taskGroup := loadLatestTaskGroup(db)
+	scheduler := NewScheduler(taskGroup, db)
+	scheduler.fillTasks(taskGroup)
+
+	return &Service{
+		config:    config,
+		db:        db,
+		scheduler: scheduler,
+	}
+}
+
+func (s *Service) Register(r *gin.RouterGroup) {
+	endpoint := r.Group("/logs")
+
+	endpoint.GET("/tasks", s.TaskGetList)
+	endpoint.GET("/tasks/:id/preview", s.TaskPreview)
+	endpoint.GET("/tasks/:id/download", s.TaskDownload)
+	endpoint.GET("/download", s.MultipleTaskDownload)
+	endpoint.POST("/tasks/retry", s.TaskRetry)
+	endpoint.POST("/tasks/cancel", s.TaskCancel)
+	endpoint.POST("/taskgroups", s.TaskGroupCreate)
+	endpoint.GET("/taskgroups/:id", s.TaskGroupGet)
+	endpoint.GET("/taskgroups/:id/preview", s.TaskGroupPreview)
+	endpoint.DELETE("/taskgroups/:id", s.TaskGroupDelete)
+}
+
+// @Summary List all tasks
+// @Description list all log search tasks
+// @Produce json
+// @Success 200 {array} TaskModel
+// @Failure 400 {object} httputil.HTTPError
+// @Router /logs/tasks [get]
+func (s *Service) TaskGetList(c *gin.Context) {
+	var tasks []TaskModel
+	err := s.db.Find(&tasks).Error
+	if err != nil {
+		httputil.NewError(c, http.StatusBadRequest, err)
+		return
+	}
+	c.JSON(http.StatusOK, tasks)
+}
+
+// @Summary Preview logs in a task
+// @Description preview fetched logs in a task by providing task ID
+// @Produce json
+// @Param id path string true "task id"
+// @Success 200 {array} PreviewModel
+// @Failure 400 {object} httputil.HTTPError
+// @Router /logs/tasks/{id}/preview [get]
+func (s *Service) TaskPreview(c *gin.Context) {
+	taskID := c.Param("id")
+	var lines []PreviewModel
+	err := s.db.Where("task_id = ?", taskID).Find(&lines).Error
+	if err != nil {
+		httputil.NewError(c, http.StatusBadRequest, err)
+		return
+	}
+	c.JSON(http.StatusOK, lines)
+}
+
+// @Summary Create a task group
+// @Description create and run a task group
+// @Produce json
+// @Success 200 {object} httputil.HTTPSuccess
+// @Failure 400 {object} httputil.HTTPError
+// @Failure 500 {object} httputil.HTTPError
+// @Router /logs/taskgroups [post]
+func (s *Service) TaskGroupCreate(c *gin.Context) {
+	// TODO: using parameters provided by client
+	endTime := time.Now().UnixNano() / int64(time.Millisecond)
+	startTime := int64(0)
+	var searchLogReq = &SearchLogRequest{
+		StartTime: startTime,
+		EndTime:   endTime,
+		Levels:    nil,
+		Patterns:  nil,
+	}
+	var components = []*Component{
+		{
+			ServerType: "tidb",
+			IP:         "127.0.0.1",
+			Port:       "4000",
+			StatusPort: "10080",
+		},
+	}
+	if s.scheduler.taskGroup != nil && s.scheduler.taskGroup.State == StateRunning {
```

Comment:
why do we need to limit this? there might be multiple pages starting a search task.